### PR TITLE
Mining borgs now have melee armor.

### PR DESCRIPTION
**Pull request by Qwertytoforty**
## What Does This PR Do

Silicons now have armor.
By default, it is zero across the board.
Mining borgs now have 20% melee armor and nothing else.
This maxes at 60% armor after 4 plates.
This armor is lost on module reset.

## Why It's Good For The Game

Considering we are on lavaland, and mining borgs kinda don't have armor when everything hits like a truck, and while miners get meds and legion cores, and all mining cyborgs get is the worlds slowest self repair system, maybe they should have melee armor.
So this pr does it.
Yes, this makes mining cyborgs *slightly* more robust for melee attacks for say, malf AI.
However, they usually die by lasers or flash into baton, which is perma stun, so this will not be an issue.

Theoretically one could now give gamma borgs armor equivlent to ERT armor, but that is out of this prs scope, and honestly should probably have a little less armor? For someone else to look at.

Armor has not been converted on syndicate borgs from flat brute / burn resist, as it would make them take more damage on emp / more damage from armor peircing rounds, but it could be done. if requested.

## Testing

Ensured melee and bullet attacks were affected by armor, as well as mob attacks, xeno attacks, and hulk attacks.
Ensured armor upgraded correctly.
Ensure armor got reset by module reset.

## Changelog
:cl:
add: Mining cyborgs now have melee armor.
add: Mining cyborgs can have there melee armor improved when someone applies golaith plates to them.
/:cl:

### CODE REVIEW

**Comment by procdrone** [comment]
Maybe this could lead into a cool upgrade for the Malf AI to give armour for all of its cyborgs? (and double for its mining borg if it gets through) (just a though/idea, not relevant to this PR)

File: code/modules/mob/living/silicon/silicon.dm [hunk 1]
```diff
@@ -211,19 +223,42 @@
...
+		var/damage = run_armor(I.force + bonus_damage, I.damtype, MELEE, 0, I.armour_penetration_flat, I.armour_penetration_percentage)
+		apply_damage(damage, I.damtype, def_zone)
+
+///returns the damage value of the attack after processing the silicons's various armor protections
+/mob/living/silicon/proc/run_armor(damage_amount, damage_type, damage_flag = 0, attack_dir, armour_penetration_flat = 0, armour_penetration_percentage = 0)
+	switch(damage_type)
+		if(BRUTE)
+		if(BURN)
+		else
+			return 0
```
Comment (Sirryan2002) [inline code/modules/mob/living/silicon/silicon.dm:254]: was there supposed to be something here?

File: code/modules/mob/living/silicon/silicon.dm [hunk 2]
```diff
@@ -211,19 +223,42 @@
...
+	var/armor_protection = 0
+	if(damage_flag)
+		armor_protection = armor.getRating(damage_flag)
+	if(armor_protection)		//Only apply weak-against-armor/hollowpoint effects if there actually IS armor.
+		armor_protection = clamp((armor_protection * ((100 - armour_penetration_percentage) / 100)) - armour_penetration_flat, min(armor_protection, 0), 100)
+	return round(damage_amount * (100 - armor_protection)*0.01, DAMAGE_PRECISION)
```
Comment (Sirryan2002) [inline code/modules/mob/living/silicon/silicon.dm:260]:
```suggestion
	return round(damage_amount * (100 - armor_protection) * 0.01, DAMAGE_PRECISION)
```

File: code/modules/mob/living/silicon/silicon_defense.dm [hunk 3]
```diff
@@ -47,7 +49,8 @@
 			to_chat(user, "<span class='warning'>You don't want to hurt [src]!</span>")
 			return FALSE
 		..(user, TRUE)
-		adjustBruteLoss(rand(10, 15))
+		var/damage = run_armor(rand(10, 15), BRUTE, MELEE)
+		adjustBruteLoss(damage)
```
Comment (Sirryan2002) [inline code/modules/mob/living/silicon/silicon_defense.dm:53]:
```suggestion
		adjustBruteLoss(run_armor(rand(10, 15), BRUTE, MELEE))
```

File: code/modules/mob/living/silicon/silicon.dm (hunk 1 above)
Comment (Qwertytoforty) [reply code/modules/mob/living/silicon/silicon.dm:254]: It's a switch, so no I think. It's checking if it's brute or burn, which will lead to damage, any other source is no damage so no need to check armor.

Comment (Sirryan2002) [reply code/modules/mob/living/silicon/silicon.dm:254]: i'd just do `if(damage_type != BRUTE && damage_type != BURN)`

**Review by Sirryan2002 — APPROVED** [review]
code looks good

**Pull request merged**